Clean up observe API

dagit opened the pull request:
This is intended to close #8 

It renames `observeMaybeT` to `observeTMaybe` to avoid confusion with `MaybeT`. Also rename `observeAllT` to `observeTAll` to be consistent.

This also changes `observe` to have type `Seq a -> Maybe a` so that the no results case is no longer an implicit exception.

@treeowl  No rush on this, but please review when you have time.

treeowl commented on src/Control/Monad/Logic/Sequence/Internal.hs:
```diff
@@ -284,25 +284,25 @@ observeT (toView -> m) = m >>= go where
 
-observeMaybeT :: Monad m => SeqT m (Maybe a) -> m (Maybe a)
-observeMaybeT (toView -> m) = m >>= go where
+observeTMaybe :: Monad m => SeqT m (Maybe a) -> m (Maybe a)
```
Is this really the right type? I would really want

```haskell
observeTMaybe :: Monad m => SeqT m a -> m (Maybe a)
```

as a way to `observeT` is a non-`MonadFail` monad. Then your version can be implemented inline quite cleanly:

```haskell
yourObserveMaybeT s = myObserveMaybeT $ s >>= maybe empty pure
```

dagit commented:
At the time I originally made that function, I had Logic (Maybe a) instead
of Logic a, to workaround observe throwing an exception when there are no
results. In hindsight maybe that should have been LogicT Maybe a. And the
only thing I wanted it for was so that I didn’t have to use something
analogous to listToMaybe . catMaybes . observeAll

But now that we can redefine the type of observe/observeT we can avoid that
implicit execution and make a nicer API, like you suggest. Does that
composition have roughly the same performance? Like does it consume all the
intermediate Nothing? I guess so due to the implicit backtracking.

On Tue, Jul 27, 2021 at 12:19 David Feuer ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/Control/Monad/Logic/Sequence/Internal.hs
> <https://github.com/dagit/logict-sequence/pull/28#discussion_r677732629>:
>
> >  {-# INLINE observe #-}
>
> -observeMaybeT :: Monad m => SeqT m (Maybe a) -> m (Maybe a)
> -observeMaybeT (toView -> m) = m >>= go where
> +observeTMaybe :: Monad m => SeqT m (Maybe a) -> m (Maybe a)
>
> Is this really the right type? I would really want
>
> observeTMaybe :: Monad m => SeqT m a -> m (Maybe a)
>
> as a way to observeT is a non-MonadFail monad. Then your version can be
> implemented inline quite cleanly:
>
> yourObserveMaybeT s = myObserveMaybeT $ s >>= maybe empty pure
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dagit/logict-sequence/pull/28#pullrequestreview-716291504>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC547JIDPZSZ5XK7I6KPUDTZ4BFXANCNFSM5BA3XLNQ>
> .
>

treeowl commented:
No idea about performance. I just think your current version feels pretty
weird from an API standpoint.

On Tue, Jul 27, 2021, 3:32 PM Jason Dagit ***@***.***> wrote:

> At the time I originally made that function, I had Logic (Maybe a) instead
> of Logic a, to workaround observe throwing an exception when there are no
> results. In hindsight maybe that should have been LogicT Maybe a. And the
> only thing I wanted it for was so that I didn’t have to use something
> analogous to listToMaybe . catMaybes . observeAll
>
> But now that we can redefine the type of observe/observeT we can avoid that
> implicit execution and make a nicer API, like you suggest. Does that
> composition have roughly the same performance? Like does it consume all the
> intermediate Nothing? I guess so due to the implicit backtracking.
>
> On Tue, Jul 27, 2021 at 12:19 David Feuer ***@***.***> wrote:
>
> > ***@***.**** commented on this pull request.
> > ------------------------------
> >
> > In src/Control/Monad/Logic/Sequence/Internal.hs
> > <https://github.com/dagit/logict-sequence/pull/28#discussion_r677732629
> >:
> >
> > > {-# INLINE observe #-}
> >
> > -observeMaybeT :: Monad m => SeqT m (Maybe a) -> m (Maybe a)
> > -observeMaybeT (toView -> m) = m >>= go where
> > +observeTMaybe :: Monad m => SeqT m (Maybe a) -> m (Maybe a)
> >
> > Is this really the right type? I would really want
> >
> > observeTMaybe :: Monad m => SeqT m a -> m (Maybe a)
> >
> > as a way to observeT is a non-MonadFail monad. Then your version can be
> > implemented inline quite cleanly:
> >
> > yourObserveMaybeT s = myObserveMaybeT $ s >>= maybe empty pure
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/dagit/logict-sequence/pull/28#pullrequestreview-716291504
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAC547JIDPZSZ5XK7I6KPUDTZ4BFXANCNFSM5BA3XLNQ
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dagit/logict-sequence/pull/28#issuecomment-887778319>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOOF7O77REBKW6FYC5P4OTTZ4CVFANCNFSM5BA3XLNQ>
> .
>

treeowl commented:
Changing the name of `observeAllT` makes it harder to use this package as a drop-in replacement for `LogicT`. I can think of two alternatives:

1. Sidestep the `MaybeT` confusion by calling it `observeMayT`.
2. Include both `observeAllT` and `observeTAll`.

dagit commented:
At one point I thought it would be nice if the observe functions were part of a type class, but that means you can't write things like this anymore:

```haskell
foo :: Int -> [a]
foo = observeAll $ do
  ....
```

You'd need some way to tell GHC which `MonadLogic` you're observing.

treeowl commented:
Writing the test suite, I have come to the conclusion that `observeT` itself is *really annoying* in the `MonadFail` era. The function I described,

```haskell
functionName :: Monad m => SeqT m a -> m (Maybe a)
```

feels much more reasonable now.

Proposal: Add a module called, say, `Control.Monad.Logic.Sequence.Legacy` with all the traditional names. Then in `Control.Monad.Logic`, use

```haskell
observeT :: Monad m => SeqT m a -> m (Maybe a)
observeAllT :: Monad m => SeqT m a -> m [a]
--etc.
```

Just a thought.

dagit commented:
I was going to suggest something similar with the modules. How about Compat?

On Wed, Jul 28, 2021 at 10:56 David Feuer ***@***.***> wrote:

> Writing the test suite, I have come to the conclusion that observeT
> itself is *really annoying* in the MonadFail era. The function I
> described,
>
> functionName :: Monad m => SeqT m a -> m (Maybe a)
>
> feels much more reasonable now.
>
> Proposal: Add a module called, say, Control.Monad.Logic.Sequence.Legacy
> with all the traditional names. Then in Control.Monad.Logic, use
>
> observeT :: Monad m => SeqT m a -> m (Maybe a)observeAllT :: Monad m => SeqT m a -> m [a]--etc.
>
> Just a thought.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dagit/logict-sequence/pull/28#issuecomment-888506776>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC547PRGTHE2SX5DA6FBODT2BAEVANCNFSM5BA3XLNQ>
> .
>

treeowl commented:
I don't understand why you just added that commit rather than rebasing. The longer you wait to rebase, the more painful it gets!

dagit commented:
Not sure what you're seeing but I did rebase.

I think this is ready for another round of review.

treeowl commented:
Oh, I got confused. I still really want to see the simple function I suggested!

dagit commented:
> 
> 
> Oh, I got confused. I still really want to see the simple function I suggested!

Hmmm...I thought I added it. Sometimes the review system gets stuck on the version you reviewed. Make sure you're looking at the last commit in this branch.

treeowl commented on src/Control/Monad/Logic/Sequence/Compat.hs:
```diff
@@ -1,5 +1,27 @@
+#endif
+observeT (toView -> m) = m >>= go where
+  go (a :< _) = return a
+  go Empty = fail "No results."
```
Shouldn't the failure message identify the function?

dagit replied:
This is for `Compat` and this is the definition that is in `LogicT`.

/shrug

treeowl commented on src/Control/Monad/Logic/Sequence/Compat.hs:
```diff
@@ -1,5 +1,27 @@
+observe :: Seq a -> a
+observe (toView -> m) = case runIdentity m of
+  a :< _ -> a
+  Empty -> error "No results."
```
Here I think the `HasCallStack` on `error` will include the function name, so maybe you don't need to do so manually?

dagit replied:
```
*Control.Monad.Logic.Sequence Control.Monad.Logic.Sequence.Compat Control.Applicative λ> Control.Monad.Logic.Sequence.Compat.observe empty
*** Exception: No results.
CallStack (from HasCallStack):
  error, called at src/Control/Monad/Logic/Sequence/Compat.hs:26:12 in logict-sequence-0.1.0.2-inplace:Control.Monad.Logic.Sequence.Compat
  ```

Looks like we're good.

treeowl reviewed (changes requested):
Mostly looks great. But in the test suite, I used `MaybeT` to hack around the lack of what you've just added, so I think the hack should be taken out. If you want to merge this first, I can take care of it after, or you can do it here.

treeowl commented:
Oh, you probably did, and I probably just read the diff wrong. Sorry about that!

dagit commented:
When you get a chance can you add more feedback or sign off on this? I think I've addressed all the feedback?

I think it would be good to merge this ahead of #41 and then rebase that one and remove the `obsFoo` stuff.

Thanks.

treeowl reviewed (approved):
Yeah, let's do this. We'll want to nix the `alot`s eventually I'm favor of an exported `choose`; you pick the order.
